### ISSUE TITLE
StringValueArray returns null if inline comment is used and some specific chars exist...

### ISSUE THREAD

**Issue by Ketan-Desai** [cemdervis/SharpConfig#102]
Hi,

Following returns StringValueArray as null

[Setting]
Columns = {A,B,C} # If quotes (") apper here

Thanks

**Comment by ignacioerrico** [comment]
I found this issue recently.

The bug is in the method `IsInQuoteMarks` of the class `ConfigurationReader`. I see it's a method that has been there since the very beginning of the project and it starts off with a [comment that basically states that it should be improved](https://github.com/cemdervis/SharpConfig/blob/master/Src/ConfigurationReader.cs#L116).

I've been going through the code and IMO trying to improve `IsInQuoteMarks` will just produce a wilder beast. My suggestion is to rethink how parsing is done.

Currently, [parsing begins by looking for the line's comment](https://github.com/cemdervis/SharpConfig/blob/master/Src/ConfigurationReader.cs#L46). I suggest a more progressive approach to parsing instead of starting with the comment.

@cemdervis, I'd be happy to look into it and submit a PR.

**Comment by cemdervis** [comment]
Hi @ignacioerrico, `IsInQuoteMarks` is indeed a naive method and parsing should be improved some time. In the beginning, the current parsing was sufficient enough (starting off with removing the comment).

We could start parsing from start to end, token by token. I assume that's what you propose. Another alternative would be to use regex matching. You are welcome to look into it, any help is appreciated.

**Comment by ignacioerrico** [comment]
Yes, that's exactly what I meant by progressive: token by token from the beginning of the string. It's actually implemented, but I would organize it in way that is more natural, which should simplify the code and at the same time prevent this kind of bugs.

By the way, I haven't found contribution guidelines. I'll just commit my changes in the `master` branch in my fork and submit a PR with that. Or do you prefer that I create a branch with a specific name (like the ticket number, for example)?

**Comment by cemdervis** [comment]
That change sounds fine. There are no special guidelines, you can just submit a PR to master and I'll integrate it after everything's checked.

**Comment by ignacioerrico** [comment]
I took some time to study the code and analyze how to improve it. After careful consideration, I discarded my original idea of parsing from beginning to end in favor of improving `ParseComment` (so, yes, I keep the logic of starting off by looking for the line's comment&mdash;**it makes sense in this context**). That resulted in the least amount of changes and each method now has a clear single responsibility. And of course, `IsInQuoteMarks` is now gone (yay!). All that is a good sign.

The new logic to look for a comment is simple: parse each line from beginning to end until you find the first valid comment character that:
1. is not within a quote (eg. `"this is # not a comment"`), **and**
2. is not escaped (eg. `this is \# not a comment either`).

Now a line like this one:

`ValidUglySetting = { first, \"second # still, second\" } # comment \"with a quote\" and a closing brace }`

is correctly parsed and produces:

Property in `section["ValidUglySetting"]` | Value
-------------------------------- | -----
`IsArray` | `true`
`ArraySize` | `2`
`StringValueArray[0]` | `first`
`StringValueArray[1]` | `second # still, second`
`Comment` | `comment \"with a quote\" and a closing brace }`

While I was at it, I cleaned up the code a bit. To ease the code review, each commit is self-contained.
